File: latest/week1/workshop2_m3/W02_M03_demo.R
# ============================================================
# Workshop 2 — Module 3: ggplot2 Deep Dive — Aesthetics & Geometries
# R Demonstration Script — UNYT Tirana
# ============================================================
library(tidyverse)
library(patchwork)
dir.create("output", showWarnings = FALSE)

# ── 1. MULTI-AESTHETIC ENCODING ───────────────────────────
set.seed(42)
df <- tibble(
  gdp = runif(80, 10, 100), life_exp = 50 + 0.3*gdp + rnorm(80, 0, 8),
  pop = runif(80, 1e6, 5e8), continent = sample(c("Africa","Asia","Europe"), 80, TRUE),
  income = sample(c("Low","Mid","High"), 80, TRUE)
)

p_bubble <- ggplot(df, aes(x = gdp, y = life_exp, color = continent,
                            size = pop, shape = income)) +
  geom_point(alpha = 0.6) +
  scale_color_manual(values = c(Africa="#1565C0", Asia="#E53935", Europe="#2E7D32")) +
  scale_size(range = c(1, 12), labels = scales::label_number(scale = 1e-6, suffix = "M")) +
  scale_shape_manual(values = c(Low=16, Mid=17, High=15)) +
  theme_minimal() +
  labs(title = "Five Variables in One Chart", x = "GDP per Capita ($K)",
       y = "Life Expectancy", size = "Population", shape = "Income")
ggsave("output/multi_aes.png", p_bubble, width = 8, height = 5, dpi = 300)

# ── 2. LAYERED GEOMS ─────────────────────────────────────
# Point + smooth
p1 <- ggplot(mpg, aes(x = displ, y = hwy)) +
  geom_point(alpha = 0.4) + geom_smooth(method = "lm", color = "#E53935") +
  theme_minimal() + labs(title = "point + smooth")

# Boxplot + jitter
p2 <- ggplot(mpg, aes(x = factor(cyl), y = hwy)) +
  geom_boxplot(outlier.shape = NA, fill = "#E3F2FD") +
  geom_jitter(width = 0.15, alpha = 0.3, size = 0.8, color = "#E53935") +
  theme_minimal() + labs(title = "boxplot + jitter", x = "Cylinders")

# Col + errorbar
mpg_summary <- mpg |> group_by(class) |>
  summarise(m = mean(hwy), se = sd(hwy)/sqrt(n()), .groups = "drop") |>
  mutate(class = fct_reorder(class, m))
p3 <- ggplot(mpg_summary, aes(x = class, y = m)) +
  geom_col(fill = "#1565C0", alpha = 0.6, width = 0.5) +
  geom_errorbar(aes(ymin = m-se, ymax = m+se), width = 0.2) +
  coord_flip() + theme_minimal() + labs(title = "col + errorbar", x = NULL, y = "Mean MPG")

p_layers <- p1 | p2 | p3
ggsave("output/geom_layers.png", p_layers, width = 12, height = 4, dpi = 300)

# ── 3. OVERPLOTTING ───────────────────────────────────────
set.seed(55)
big <- tibble(x = rnorm(5000, 50, 15), y = 0.3*x + rnorm(5000, 0, 8))

p_over <- ggplot(big, aes(x, y)) + geom_point(size = 0.5) +
  theme_minimal() + labs(title = "Overplotting (n=5000)")
p_alpha <- ggplot(big, aes(x, y)) + geom_point(alpha = 0.05, size = 0.5) +
  theme_minimal() + labs(title = "Fix: alpha = 0.05")
p_hex <- ggplot(big, aes(x, y)) + geom_hex(bins = 25) +
  scale_fill_viridis_c() + theme_minimal() + labs(title = "Fix: geom_hex()")
p_dens <- ggplot(big, aes(x, y)) + geom_density_2d_filled(alpha = 0.7) +
  theme_minimal() + labs(title = "Fix: geom_density_2d_filled()")

p_overplot <- (p_over | p_alpha) / (p_hex | p_dens) +
  plot_annotation(title = "Overplotting: Problem and Solutions")
ggsave("output/overplotting.png", p_overplot, width = 10, height = 7, dpi = 300)

# ── 4. LOLLIPOP CHART ────────────────────────────────────
p_lollipop <- mpg_summary |>
  ggplot(aes(x = class, y = m)) +
  geom_segment(aes(xend = class, y = 0, yend = m), color = "grey70") +
  geom_point(size = 3, color = "#1565C0") +
  coord_flip() + theme_minimal() +
  labs(title = "Lollipop: geom_segment + geom_point", x = NULL, y = "Mean Highway MPG")
ggsave("output/lollipop.png", p_lollipop, width = 6, height = 4, dpi = 300)

# ── 5. DUMBBELL CHART ────────────────────────────────────
df_db <- tibble(
  cat = c("Economy","Midsize","Compact","SUV","Pickup"),
  y2008 = c(26, 24, 28, 18, 16), y2024 = c(32, 30, 35, 24, 22)
) |> mutate(cat = fct_reorder(cat, y2024))

p_dumbbell <- ggplot(df_db, aes(y = cat)) +
  geom_segment(aes(x = y2008, xend = y2024, yend = cat), color = "grey80", linewidth = 1.5) +
  geom_point(aes(x = y2008), size = 3, color = "#1565C0") +
  geom_point(aes(x = y2024), size = 3, color = "#E53935") +
  theme_minimal() +
  labs(title = "Dumbbell: 2008 (blue) vs 2024 (red)", x = "MPG", y = NULL)
ggsave("output/dumbbell.png", p_dumbbell, width = 7, height = 4, dpi = 300)

cat("\n── All W02-M03 R plots saved ──\n")

File: latest/week1/workshop2_m3/W02_M03_demo.py
"""W02-M03: ggplot2 Deep Dive — Python Script — UNYT Tirana"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
os.makedirs("output", exist_ok=True); np.random.seed(42)

# 1. MULTI-AESTHETIC (bubble chart)
n=80; gdp=np.random.uniform(10,100,n); le=50+0.3*gdp+np.random.normal(0,8,n)
pop=np.random.uniform(1e6,5e8,n); cont=np.random.choice(["Africa","Asia","Europe"],n)
pal={"Africa":"#1565C0","Asia":"#E53935","Europe":"#2E7D32"}
fig,ax=plt.subplots(figsize=(7,5))
for g,c in pal.items():
    m=cont==g; ax.scatter(gdp[m],le[m],s=pop[m]/3e6,c=c,alpha=0.5,edgecolors="white",linewidth=0.5,label=g)
ax.legend(fontsize=7,title="Continent"); ax.set_xlabel("GDP ($K)"); ax.set_ylabel("Life Expectancy")
ax.set_title("Bubble: 4 variables (x, y, color, size)",fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/multi_aes_py.png",dpi=300); plt.close()

# 2. OVERPLOTTING — 4 panels
np.random.seed(55); n2=2000; x=np.random.normal(50,15,n2); y=0.3*x+np.random.normal(0,8,n2)
fig,axes=plt.subplots(1,4,figsize=(14,3.5))
axes[0].scatter(x,y,s=3,c="#1565C0"); axes[0].set_title("Overplotting",fontsize=8,fontweight="bold",color="#C62828")
axes[1].scatter(x,y,s=3,c="#1565C0",alpha=0.05); axes[1].set_title("alpha=0.05",fontsize=8,fontweight="bold")
axes[2].scatter(x+np.random.normal(0,1,n2),y+np.random.normal(0,1,n2),s=3,c="#1565C0",alpha=0.08)
axes[2].set_title("jitter + alpha",fontsize=8,fontweight="bold")
axes[3].hexbin(x,y,gridsize=20,cmap="YlGnBu",mincnt=1); axes[3].set_title("hexbin",fontsize=8,fontweight="bold")
for ax in axes: ax.tick_params(labelsize=5); ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Overplotting Solutions",fontsize=11,fontweight="bold"); plt.tight_layout()
plt.savefig("output/overplotting_py.png",dpi=300); plt.close()

# 3. LOLLIPOP
cats=["Compact","Midsize","SUV","Pickup","Minivan"]; vals=[28,24,18,16,22]
idx=np.argsort(vals); sc=[cats[i] for i in idx]; sv=[vals[i] for i in idx]
fig,ax=plt.subplots(figsize=(6,4))
ax.hlines(y=range(len(sc)),xmin=0,xmax=sv,color="#BBBBBB",linewidth=1.5)
ax.scatter(sv,range(len(sc)),s=60,c="#1565C0",zorder=5,edgecolors="white",linewidth=0.5)
ax.set_yticks(range(len(sc))); ax.set_yticklabels(sc)
ax.set_title("Lollipop Chart",fontweight="bold"); ax.set_xlabel("Mean MPG")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/lollipop_py.png",dpi=300); plt.close()

# 4. DUMBBELL
cats2=["Economy","Midsize","Compact","SUV","Pickup"]; v2008=[26,24,28,18,16]; v2024=[32,30,35,24,22]
idx2=np.argsort(v2024); sc2=[cats2[i] for i in idx2]; s08=[v2008[i] for i in idx2]; s24=[v2024[i] for i in idx2]
fig,ax=plt.subplots(figsize=(7,4))
for i,(c,a,b) in enumerate(zip(sc2,s08,s24)):
    ax.plot([a,b],[i,i],color="#CCCCCC",linewidth=2,solid_capstyle="round")
    ax.scatter([a],[i],s=60,c="#1565C0",zorder=5,edgecolors="white",linewidth=0.5)
    ax.scatter([b],[i],s=60,c="#E53935",zorder=5,edgecolors="white",linewidth=0.5)
ax.set_yticks(range(len(sc2))); ax.set_yticklabels(sc2)
ax.scatter([],[],s=40,c="#1565C0",label="2008"); ax.scatter([],[],s=40,c="#E53935",label="2024")
ax.legend(fontsize=7); ax.set_title("Dumbbell: 2008 vs 2024",fontweight="bold"); ax.set_xlabel("MPG")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/dumbbell_py.png",dpi=300); plt.close()

print("All W02-M03 Python plots saved")
